sycl/doc/design/spirv-extensions/SPV_INTEL_masked_gather_scatter.asciidoc

```diff
@@ -0,0 +1,288 @@
+|========================================
+| Last Modified Date | 2022-08-19
+| Revision           | A
+|========================================
+
+== Dependencies
+
+This extension is written against the SPIR-V Specification,
+Version 1.5 Revision 5.
+
+This extension requires SPIR-V 1.0.
+
+== Overview
+
+This extension allows *TypeVector* to have a <<Physical Pointer Type>> _Component Type_ and introduces gather/scatter instructions.
+It will be useful for explicitly vectorized kernels.
+
+== Extension Name
+
+
+To use this extension within a SPIR-V module, the appropriate *OpExtension* must
+be present in the module:
+
+[subs="attributes"]
+----
+OpExtension "{extension_name}"
+----
+
+== New Capabilities
+
+This extension introduces new capabilities:
+
+[subs="attributes"]
+----
+{capability_name}
+----
+
+== New Instructions
+
+Instructions added under the *{capability_name}* capability:
+
+----
+OpMaskedGatherINTEL
+OpMaskedScatterINTEL
+----
+
+== Token Number Assignments
+
+[width="40%"]
+[cols="70%,30%"]
+[grid="rows"]
+|====
+|*{capability_name}*  | {capability_token}
+|*OpMaskedGatherINTEL*    | {OpMaskedGatherINTEL_token}
+|*OpMaskedScatterINTEL*    | {OpMaskedScatterINTEL_token}
+|====
+
+== Modifications to the SPIR-V Specification, Version 1.5
+
+
+=== 2.16.1. Universal Validation Rules
+
+Modify Data rules section, replacing following segment:
+
+- Vector types must be parameterized only with numerical types or the <<OpTypeBool>> type.
+
+with:
+
+- Vector types must be parameterized only with numerical types or the <<OpTypeBool>> type. They can also
+ be parameterized with <<Physical Pointer Type>> types under *{capability_name}* capability.
+
+=== 2.2.2. Types
+
+Define _Vector_ as:
+An ordered homogeneous collection of two or more <<scalars>> or pointers of <<Physical Pointer Type>>.
+Vector sizes are quite restrictive and dependent on the execution model.
+
+
+=== Capabilities
+
+Modify Section 3.31, Capability, adding rows to the Capability table:
+
+--
+[options="header"]
+|====
+2+^| Capability ^| Implicitly Declares 
+| {capability_token}
+| *{capability_name}* +
+ +
+Allow *TypeVector* to have a <<Physical Pointer Type,Physical Pointer Type>> _Component Type_. +
+ +
+See also extension: *{extension_name}* +
+|
+*Kernel*
+|====
+--
+
+
+=== 3.42.6. Type-Declaration Instructions
+
+Modify <<OpTypeVector>>, expounding _Component Type_ as:
+  Component Type is the type of each component in the resulting type. It must be a <<scalar type>> or <<Physical Pointer Type>>.
+
+
+=== 3.42.8. Memory Instructions
+
+Allow <<vector>> with <<physical pointer type>> to be used by <<OpVariable>>, <<OpAccessChain>>, <<OpInBoundsAccessChain>>,
+<<OpPtrAccessChain>>, <<OpInBoundsPtrAccessChain>>, <<OpPtrEqual>>, <<OpPtrNotEqual>> and <<OpPtrDiff>> instructions. When <<vector>> with
+<<physical pointer type>> is allowed for <<OpVariable>> it is implicitly possible to be used by <<OpStore>> and <<OpLoad>> which can
```

Comment:
I might not be reading this correctly? But I thought vector of pointers will be used by OpMaskedGatherINTEL and OpMaskedScatterINTEL?

Reply:
`OpStore` can take the vector of pointer as an Object input
`OpLoad` can result in a vector of pointers, taking a pointer to a vector of pointers as an input vector.
So it's not like we can store a vector of integers to a vector of pointers to integers.

I see, that these lines might be a bit confusing, but I felt we need to add them, because we modify `OpVariable` definition and it itself has a following line:
"Allocate an object in memory, resulting in a pointer to it, which can be used with OpLoad and OpStore.".

It would be great, if you suggest a better wording.